osquery/tables/system/windows/windows_crashes.cpp

```diff
@@ -0,0 +1,764 @@
+    {0x00000200, "MiniDumpWithPrivateReadWriteMemory"},
+    {0x00000400, "MiniDumpWithoutOptionalData"},
+    {0x00000800, "MiniDumpWithFullMemoryInfo"},
+    {0x00001000, "MiniDumpWithThreadInfo"},
+    {0x00002000, "MiniDumpWithCodeSegs"},
+    {0x00004000, "MiniDumpWithoutAuxiliaryState"},
+    {0x00008000, "MiniDumpWithFullAuxiliaryState"},
+    {0x00010000, "MiniDumpWithPrivateWriteCopyMemory"},
+    {0x00020000, "MiniDumpIgnoreInaccessibleMemory"},
+    {0x00040000, "MiniDumpWithTokenInformation"},
+    {0x00080000, "MiniDumpWithModuleHeaders"},
+    {0x00100000, "MiniDumpFilterTriage"},
+    {0x001fffff, "MiniDumpValidTypeFlags"}};
+const std::vector<MINIDUMP_STREAM_TYPE> kStreamTypes = {ExceptionStream,
+                                                        ModuleListStream,
+                                                        ThreadListStream,
+                                                        MemoryListStream,
+                                                        SystemInfoStream,
+                                                        MiscInfoStream};
+
+Status logAndStoreTID(const MINIDUMP_EXCEPTION_STREAM* stream,
+                      unsigned int& tid,
+                      Row& r) {
+  if ((stream == nullptr) || (stream->ThreadId == 0))
+    return Status(1);
+
+  r["tid"] = BIGINT(stream->ThreadId);
+  tid = stream->ThreadId;
+  return Status(0);
+}
+
+Status logAndStoreExceptionCodeAndAddr(const MINIDUMP_EXCEPTION_STREAM* stream,
+                                       unsigned int& exCode,
+                                       unsigned long long& exAddr,
+                                       Row& r) {
+  if (stream == nullptr)
+    return Status(1);
+
+  auto ex = stream->ExceptionRecord;
+  std::ostringstream exCodeStr;
+  exCodeStr << "0x" << std::hex << ex.ExceptionCode;
+  if (exCodeStr.fail())
+    return Status(1);
+  r["exception_code"] = exCodeStr.str();
+  exCode = ex.ExceptionCode;
+
+  std::ostringstream exAddrStr;
+  exAddrStr << "0x" << std::hex << ex.ExceptionAddress;
+  if (exAddrStr.fail())
+    return Status(1);
+  r["exception_address"] = exAddrStr.str();
+  exAddr = ex.ExceptionAddress;
+  return Status(0);
+}
+
+/*
+Log the exception message for errors with defined parameters
+(see ExceptionInformation @
+https://msdn.microsoft.com/en-us/library/windows/desktop/ms680367(v=vs.85).aspx)
+*/
+Status logExceptionMsg(const MINIDUMP_EXCEPTION_STREAM* stream,
+                       unsigned int exCode,
+                       unsigned long long exAddr,
+                       Row& r) {
+  if (stream == nullptr)
+    return Status(1);
+
+  auto ex = stream->ExceptionRecord;
+  if ((ex.ExceptionCode == EXCEPTION_ACCESS_VIOLATION) &&
+      (ex.NumberParameters == 2)) {
+    std::ostringstream errorMsg;
+    std::ostringstream memAddrStr;
+    memAddrStr << "0x" << std::hex << ex.ExceptionInformation[1];
+    std::ostringstream exAddrStr;
+    exAddrStr << "0x" << std::hex << exAddr;
```

Comment:
I think this is unnecessary, considering the `exAddr` string is already in it's own column, bringing it into the `exception_message` seems excessive. I'd say perhaps here, it's sufficient to just report the memory address and the issue, and this might even further drop your requirement for a `stringstream`. Perhaps something like this?
```
std::string error {""};
switch (ex.ExceptionInformation[0]) {
case 0:
  error = "Memory at " + exAddr.str() + " could not be read.";
  break;
case 1:
  error = "Memory at " + exAddr.str() + " could not be written.";
  break;
case 8:
  error = "Memory at " + exAddr.str() + " encountered a DEP violation.";
  break;
default:
  break;
}
r["exception_message"] = error.empty() ? "unknown" : error;
return Status();
```

Reply:
I was going for reproducing the NTSTATUS (https://msdn.microsoft.com/en-us/library/cc704588.aspx) message for STATUS_ACCESS_VIOLATION: "The instruction at 0x%08lx referenced memory at 0x%08lx. The memory could not be %s.". The message for STATUS_IN_PAGE_ERROR is similar: "The instruction at 0x%08lx referenced memory at 0x%08lx. The required data was not placed into memory because of an I/O error status of 0x%08lx."

But what you said makes sense, I'd be game for doing our own thing and simplifying the message down. Your call: MS way or our way?

Reply:
Actually after some more thought, I really don't like the NTSTATUS error message for STATUS_IN_PAGE_ERROR.. I'll change these around to the format you have above

Reply:
I would say ours. Let's shoot for simple data representations inside of columns. It can sometimes be a pain for folks to look up data on their side to discern what error codes mean, but the thing I try to keep in mind is many osquery deployments can be **massive**, and sometimes it's worth making tradeoffs and encouraging folks to do data processing on the backend. I'd say for this table in particular, we're going to want to try and be terse where possible, as we're offering up a lot of columns, and we're already running the risk of presenting too much data :)